_posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md

```diff
@@ -0,0 +1,71 @@
+---
+title: 'Only use named scopes outside models'
+description: 'Keep your wheres to yourself'
+layout: article
+category: ruby
+image:
+  base: '2018/only-use-named-scopes-outside-models'
+  alt: 'San Francisco Museum of Modern Art'
+  source: 'https://unsplash.com/photos/RkJF2BMrLJc'
+  credit: 'William Bout'
+---
+
+Last time we discussed using the hash-syntax in your `#where` methods but in my examples I did something I wouldn't do in my real-life code...
+
+## Instead of…
+
+...using `#where` scopes in your controllers or views.
+
+```ruby
+class PostsController < ApplicationController
+  def index
+    @posts = Post.where(status: 'published')
+  end
+end
+```
+
+
+## Use…
+
+...only named scopes that you define in your model.
+
+```ruby
+class Post < ApplicationRecord
+  scope :published, -> { where(status: 'published') }
+end
+```
+
+And then use it like so:
+
+```ruby
+class PostsController < ApplicationController
+  def index
+    @posts = Post.published
+  end
+end
+```
+
+
+## But why?
+
+This technique improves the organisation of your code. It forces you to do two things that can really help keep you productive over time.
+
+Firstly: naming the concepts you're creating. When you name something, you explain it, often for the benefit of “future you” or your colleagues.
+
+Secondly: you have _one_ place to look for all of this. If you define scopes outside your models you can end up with arbitrary scopes and conditions defined all over your code base. When you know where all the conditions are defined you'll know where to look when you want to refactor or optimise database performance.
+
+
+## Why not?
+
+For scopes involving `#limit`, simple `#order`s or pagination (if you're doing that) there's very little point in bothering to create specific scopes.
+
+Often, wrapping non-`#where` queries does not enhance understanding. There’s no extra clarity from wrapping simple `ActiveRelation` methods inside a scope.
+
+```ruby
+scope :by_title, -> { sort(:title) } # no benefit?
+scope :recently_updated, -> { sort(updated_at: :desc) } # worth doing
```

Comment:
Hmmm...aren't both of these examples of 'wrapping simple `ActiveRelation` methods inside a scope'?

If you'd named the scope on like 65 as :by_updated_at you could argue that that is also not worth doing, right?

So, I'm not convinced by the paragraph at line 62 (which still seems like saying the same thing twice) and then the examples. I'm not clear on what's "simple" or not..as the thing on line 66 seems simple to me.

Reply:
They are, but one is an example of a 'decent name' and one a 'probably pointless' name. If the name was `by_updated_at` I'd say it was terrible name.

The naming helps explain when you use it. The ordering of `updated_at` means you need extra context, is it 'recently changed' or 'least recently'.

Reply:
Got it...but maybe you need to reword the preceding paragraph to stress that naming is only beneficial when it helps enhance the meaning...maybe spell out why sort(:title) is pointless (no extra information) sorting by :updated_at needs more clarity.